### PR TITLE
feat(protocol): Batch Encoding

### PR DESCRIPTION
**Pull request by refcell**
### Description

Implements encoding for batch types.

Steps in the direction to batch submission.

### CODE REVIEW

**Comment by refcell** [comment]
## 📚 $\text{Stack Overview}$

Pulls submitted in this stack:
* #259 👈
* `main`

_This comment was automatically generated by [`st`](https://github.com/clabby/st)._

File: crates/protocol/src/batch/core.rs [hunk 1]
```diff
@@ -48,4 +51,82 @@ impl Batch {
             }
         }
     }
+
+    /// Attempts to encode the batch to a writer.
+    pub fn encode(&self, out: &mut Vec<u8>) -> Result<(), BatchEncodingError> {
```
Comment (mattsse) [inline crates/protocol/src/batch/core.rs:56]: you could also use `out: &mut dyn bytes::BufMut` here then this accepts vec and bytesmut
but vec is fine

File: crates/protocol/src/batch/raw.rs [hunk 2]
```diff
@@ -16,10 +16,10 @@ pub struct RawSpanBatch {
     pub payload: SpanBatchPayload,
 }
 
-impl TryFrom<SpanBatch> for RawSpanBatch {
+impl TryFrom<&SpanBatch> for RawSpanBatch {
```
Comment (mattsse) [inline crates/protocol/src/batch/raw.rs:19]: TryFrom references are usually horrbile to work with, recommend adding function to`SpanBatch` as well `try_to_raw(&self` or smth

**Review by mattsse — APPROVED** [review]

File: crates/protocol/src/batch/raw.rs (hunk 2 above)
Comment (refcell) [reply crates/protocol/src/batch/raw.rs:19]: Yea good callout. Will follow up with this!

**Pull request merged**